src/charm.py

```diff
@@ -561,29 +582,101 @@ def _on_mongod_pebble_ready(self, event) -> None:
         container.exec(["chmod", "644", "/etc/logrotate.d/mongodb"])
 
-        # Restart changed services and start startup-enabled services.
-        container.replan()
+        return modified
+
+    def _configure_container(self, container: Container):
+        """Configure MongoDB pebble layer specification."""
+        if not container.can_connect():
+            logger.debug("mongod container is not ready yet.")
+            raise ContainerNotReadyError
+
+        # We need to check that the storages are attached before starting the services.
+        # pebble-ready is not guaranteed to run after storage-attached so this check allows
+        # to ensure that the storages are attached before the pebble-ready hook is run.
+        if any(not storage for storage in self.model.storages.values()):
+            logger.debug("Storages are not attached yet")
+            raise ContainerNotReadyError
+
+        if not self.__filesystem_handler(container):
+            raise ContainerNotReadyError
+
+        if self.__configure_layers(container):
+            container.replan()
 
         # when a network cuts and the pod restarts - reconnect to the exporter
         try:
             self._connect_mongodb_exporter()
             self._connect_pbm_agent()
         except MissingSecretError as e:
             logger.error("Cannot connect mongodb exporter: %r", e)
+            raise ContainerNotReadyError
+
+    # BEGIN: charm events
+    def _on_upgrade(self, event: UpgradeCharmEvent):
+        """Upgrade event handler.
+
+        During an upgrade event, it will set the version in all relations,
+        replan the container and process the upgrade statuses. If the upgrade
+        is compatible, it will end up emitting a post upgrade event that
+        verifies the health of the cluster.
+        """
+        if self.unit.is_leader():
+            self.version_checker.set_version_across_all_relations()
+
+        container = self.unit.get_container(Config.CONTAINER_NAME)
+
+        # Just run the configure layers steps on the container and defer if it fails.
+        try:
+            self._configure_container(container)
+        except ContainerNotReadyError:
             event.defer()
             return
 
+        self.status.set_and_share_status(ActiveStatus())
+        self.upgrade._reconcile_upgrade(event)
+        if self.upgrade._upgrade.is_compatible:
+            # Post upgrade event verifies the success of the upgrade.
+            self.upgrade.post_app_upgrade_event.emit()
+
+    def _on_mongod_pebble_ready(self, event) -> None:
+        """Configure MongoDB pebble layer specification."""
+        container = self.unit.get_container(Config.CONTAINER_NAME)
+
+        # Just run the configure layers steps on the container and defer if it fails.
+        try:
+            self._configure_container(container)
+        except ContainerNotReadyError:
+            event.defer()
+            return
```

Comment:
I could be wrong, but I think if the container doesn't start, that juju keeps scheduling this event so there isn't a need to defer?

Reply:
Hum maybe but what I saw in the initial training, I was told to defer in those cases, so maybe there's something that I missed?